src/langmem/short_term/summarization.py

```diff
@@ -0,0 +1,357 @@
+from dataclasses import dataclass
+from typing import Any, Callable, cast
+
+from langchain_core.language_models.chat_models import BaseChatModel
+from langchain_core.messages import AIMessage, AnyMessage, HumanMessage, SystemMessage
+from langchain_core.prompts.chat import ChatPromptTemplate, ChatPromptValue
+from langgraph.utils.runnable import RunnableCallable
+from pydantic import BaseModel
+
+TokenCounter = Callable[[list[AnyMessage]], int]
+
+
+DEFAULT_INITIAL_SUMMARY_PROMPT = ChatPromptTemplate.from_messages(
+    [
+        ("placeholder", "{messages}"),
+        ("user", "Create a summary of the conversation above:"),
+    ]
+)
+
+
+DEFAULT_EXISTING_SUMMARY_PROMPT = ChatPromptTemplate.from_messages(
+    [
+        ("placeholder", "{messages}"),
+        (
+            "user",
+            "This is summary of the conversation so far: {existing_summary}\n\n"
+            "Extend this summary by taking into account the new messages above:",
+        ),
+    ]
+)
+
+DEFAULT_FINAL_SUMMARY_PROMPT = ChatPromptTemplate.from_messages(
+    [
+        # if exists
+        ("placeholder", "{system_message}"),
+        ("system", "Summary of the conversation so far: {summary}"),
+        ("placeholder", "{messages}"),
+    ]
+)
+
+
+@dataclass
+class RunningSummary:
+    """Object for storing information about the previous summarization.
+
+    Used on subsequent calls to summarize_messages to avoid summarizing the same messages.
+    """
+
+    summary: str
+    """Latest summary of the messages, updated every time the summarization is performed."""
+
+    summarized_message_ids: set[str]
+    """The IDs of all of the messages that have been previously summarized."""
+
+
+@dataclass
+class SummarizationResult:
+    """Result of message summarization."""
+
+    messages: list[AnyMessage]
+    """List of updated messages that are ready to be input to the LLM after summarization, including a message with a summary (if any)."""
+
+    running_summary: RunningSummary | None = None
+    """Information about previous summarization (the summary and the IDs of the previously summarized messages.
+    Can be None if no summarization was performed (not enough messages to summarize).
+    """
+
+
+def summarize_messages(
+    messages: list[AnyMessage],
+    *,
+    running_summary: RunningSummary | None,
+    model: BaseChatModel,
+    max_tokens: int,
+    max_summary_tokens: int = 1,
```

Comment:
Why default to 1?

Reply:
yes! will be updated w/ core change (it was a stupid temp default while default `token_counter` was `len`)